### PR TITLE
GroupArithmetic#2630

### PR DESCRIPTION
**Pull request by Natrezim**
- member object now has sourceGroupId parameter. It defines from which
  group is member propagated.
- group union operation implemented
- tests implemented

### CODE REVIEW

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/Member.java [hunk 1]
```diff
@@ -126,15 +141,16 @@ public String serializeToString() {
...
+				", sourceGroupId=<").append(getSourceGroupId() == null ? "\\0" : BeansUtils.createEscaping(getSourceGroupId().toString())).append(">").append(
 			']').toString();
 	}
 
 	@Override
 	public String toString() {
 		StringBuilder str = new StringBuilder();
-
-		return str.append("Member:[id='").append(getId()).append("', userId='").append(userId).append("', voId='").append(voId).append("', status='").append(status).append("', type='").append(membershipType).append("']").toString();
+
```
Comment (stavamichal) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/Member.java]: Remove this space please.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/GroupsManagerBl.java [hunk 2]
```diff
@@ -1035,4 +1034,73 @@
...
+	 * 3) group removal
+	 *
+	 * @param sess perun session
+	 * @param resultGroup group to which members are added or removed from
+	 * @param changedMembers list of changed members which is passed as argument to add/remove indirect/excluded members methods. 
+	 *                          Based on these four types of operations the list contains:
+	 *                       	1) adding indirect (<code>addition</code> true) - records of added indirect members from operand group.
+	 *                       	2) removing indirect (<code>addition</code> false) - records of removed indirect members from operand group.
+	 *                       	3) adding excluded (<code>addition</code> false) - records of added excluded members from operand group.
+	 *                       	4) removing excluded (<code>addition</code> true) -  records of removed excluded members from operand group.
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/bl/GroupsManagerBl.java]: Javadoc.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 3]
```diff
@@ -176,26 +187,24 @@ protected void deleteAnyGroup(PerunSession sess, Group group, boolean forceDelet
...
-
-				Integer parentGroupId = g.getParentGroupId();
-				while(parentGroupId != null) {
-					Group parentGroup;
-					try {
-						parentGroup = getGroupById(sess, parentGroupId);
-					} catch (GroupNotExistsException ex) {
-						throw new ConsistencyErrorException(ex);
-					}
-					List<Member> membersFromParentGroup = getGroupMembers(sess, parentGroup);
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Please add comment. There are actual members of parentGroup after removing members from subGroup.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 4]
```diff
@@ -258,11 +267,29 @@ protected void deleteAnyGroup(PerunSession sess, Group group, boolean forceDelet
...
+		// Deletes also all direct and indirect members of the group
 		getGroupsManagerImpl().deleteGroup(sess, vo, group);
 
-		Integer parentGroupId = group.getParentGroupId();
+		logAuditer(sess, group.getParentGroupId(), membersFromDeletedGroup);
+
+		getPerunBl().getAuditer().log(sess, "{} deleted.", group);
+	}
+
+	private void logAuditer(PerunSession sess, Integer parentGroupId, List<Member> membersFromDeletedGroup) throws InternalErrorException {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Rename this method to "logTotallyRemovedMembers" + javadoc.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 5]
```diff
@@ -277,11 +304,9 @@ protected void deleteAnyGroup(PerunSession sess, Group group, boolean forceDelet
 			}
 			parentGroupId=parentGroup.getParentGroupId();
 		}
-
-		getPerunBl().getAuditer().log(sess, "{} deleted.", group);
 	}
-
-	public void deleteAllGroups(PerunSession sess, Vo vo) throws InternalErrorException, GroupAlreadyRemovedException, GroupAlreadyRemovedFromResourceException {
+
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Remove space.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 6]
```diff
@@ -388,35 +413,80 @@ public void addMember(PerunSession sess, Group group, Member member) throws Inte
...
 		member = getGroupsManagerImpl().addMember(sess, group, member, MembershipType.DIRECT, group.getId());
 		getPerunBl().getAuditer().log(sess, "{} added to {}.", member, group);
 
-		// only if group have parent !!
-		if (group.getParentGroupId() != null) {
-			for(Group parentGroups : getParentGroups(sess, group)){
-				member = getGroupsManagerImpl().addMember(sess, parentGroups, member, MembershipType.INDIRECT, group.getId());
-				getPerunBl().getAuditer().log(sess, "{} added to {}.", member, parentGroups);
-			}
+		setRequiredAttributes(sess, member, group);
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Please move this line to the end of this method.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 7]
```diff
@@ -388,35 +413,80 @@ public void addMember(PerunSession sess, Group group, Member member) throws Inte
...
+	 * @throws WrongAttributeValueException
+	 * @throws WrongReferenceAttributeValueException
+	 */
+	protected List<Member> addIndirectMembers(PerunSession sess, Group group, List<Member> members, int sourceGroupId) throws InternalErrorException, AlreadyMemberException, WrongAttributeValueException, WrongReferenceAttributeValueException {
+		// save list of old group members
+		List<Member> oldMembers = this.getGroupMembers(sess, group);
+
+		for (Member member : members) {
+			groupsManagerImpl.addMember(sess, group, member, MembershipType.INDIRECT, sourceGroupId);
+			setRequiredAttributes(sess, member, group);
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Move this "setRequiredAttributes" before loging in newMembers cycle.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 8]
```diff
@@ -424,35 +494,64 @@ protected void addMemberInternal(PerunSession sess, Group group, Member member)
...
 		} else {
 			throw new InternalErrorException("This method must be called only from methods VosManager.removeAdmin and MembersManager.deleteMember.");
 		}
 	}
 
-	protected void removeMemberInternal(PerunSession sess, Group group, Member member) throws InternalErrorException, NotGroupMemberException {
-
-		member.setMembershipType(MembershipType.DIRECT);
+	protected void removeDirectMember(PerunSession sess, Group group, Member member) throws InternalErrorException, NotGroupMemberException, GroupNotExistsException, GroupOperationsException {
+
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Space.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 9]
```diff
@@ -1906,6 +2013,8 @@ private void addMissingMembersWhileSynchronization(PerunSession sess, Group grou
 					} catch(NotMemberOfParentGroupException ex) {
 						// Shouldn't happen, because every group has at least Members group as a parent
 						throw new ConsistencyErrorException(ex);
+					} catch (GroupNotExistsException e) {
+						e.printStackTrace();
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: ConsistencyErrorException.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 10]
```diff
@@ -2045,4 +2154,105 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+		if (parentFlag || groupsManagerImpl.isRelationRemovable(sess, resultGroup, operandGroup)) {
+			processRelationMembers(sess, resultGroup, getGroupMembers(sess, operandGroup), operandGroup.getId(), false);	
+		} else {
+			throw new GroupOperationsException("This union cannot be removed, because it's part of the hierarchical structure of the groups.");
+		}
+		
+		groupsManagerImpl.removeGroupUnion(sess, resultGroup, operandGroup);
+	}
+
+	public boolean checkGroupsCycle(PerunSession sess, int resultGroupId, int operandGroupId) throws InternalErrorException {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Check if any of those method below can be private.

File: perun-core/src/main/java/cz/metacentrum/perun/core/entry/GroupsManagerEntry.java [hunk 11]
```diff
@@ -992,4 +992,43 @@ public RichGroup getRichGroupByIdWithAttributesByNames(PerunSession sess, int gr
...
+		
+		return getGroupsManagerBl().createGroupUnion(sess, resultGroup, operandGroup, false);
+	}
+
+	@Override
+	public void removeGroupUnion(PerunSession sess, Group resultGroup, Group operandGroup) throws InternalErrorException, GroupNotExistsException, PrivilegeException, GroupOperationsException {
+		Utils.checkPerunSession(sess);
+		getGroupsManagerBl().checkGroupExists(sess, resultGroup);
+		getGroupsManagerBl().checkGroupExists(sess, operandGroup);
+		if (!getGroupsManagerBl().isOneWayRelationBetweenGroups(resultGroup, operandGroup)) {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/entry/GroupsManagerEntry.java]: Check if this test of relation is also in BlImpl part. If not, move it there.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java [hunk 12]
```diff
@@ -279,8 +281,9 @@ public boolean isUserMemberOfGroup(PerunSession sess, User user, Group group) th
 
 	public List<Member> getGroupMembers(PerunSession sess, Group group) throws InternalErrorException {
 		try {
-			return jdbc.query("select " + MembersManagerImpl.memberMappingSelectQuery + ", groups_members.membership_type as membership_type from groups_members join members on members.id=groups_members.member_id " +
-					" where groups_members.group_id=?", MembersManagerImpl.MEMBER_MAPPER, group.getId());
+			return jdbc.query("select " + MembersManagerImpl.memberMappingSelectQuery + ", groups_members.membership_type as membership_type,  " +
+					"groups_members.source_group_id as source_group_id from groups_members join members on members.id=groups_members.member_id " +
+					"where groups_members.group_id=?", MembersManagerImpl.MEMBER_MAPPER, group.getId());
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java]: Line 291/294 getGroupMembers method, please rename parameter excludeStatusInsteadOfIncludeStatus to something better.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java [hunk 13]
```diff
@@ -483,7 +489,7 @@ public boolean isDirectGroupMember(PerunSession sess, Group group, Member member
 	public void removeMember(PerunSession sess, Group group, Member member) throws InternalErrorException, NotGroupMemberException {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java]: Please add to javadoc information about NEEDED sourceGroupId.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/MembersManagerImpl.java [hunk 14]
```diff
@@ -49,9 +49,10 @@ public Member mapRow(ResultSet rs, int i) throws SQLException {
 					rs.getString("members_created_at"), rs.getString("members_created_by"), rs.getString("members_modified_at"), rs.getString("members_modified_by"),
 					rs.getInt("members_created_by_uid") == 0 ? null : rs.getInt("members_created_by_uid"),
 					rs.getInt("members_modified_by_uid") == 0 ? null : rs.getInt("members_modified_by_uid"));
-			try{
+			try {
+				member.setSourceGroupId(rs.getInt("source_group_id"));
 				member.setMembershipType(MembershipType.getMembershipType(rs.getInt("membership_type")));
-			}catch(SQLException ex){/*member.setType(MembershipType.NOT_DEFINED); již provedeno v konstruktoru*/}
+			} catch(SQLException ex){/*member.setType(MembershipType.NOT_DEFINED); již provedeno v konstruktoru*/}
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/MembersManagerImpl.java]: Please fix the text to English.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java (hunk 13 above)
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java]: Check on sourceGroupId != null is there, but there is still missing information in javadoc, that member with set sourceGroupId is needed.

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/exceptions/GroupOperationsException.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+package cz.metacentrum.perun.core.api.exceptions;
+
+/**
+ * Class description.
```
Comment (licehammer) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/exceptions/GroupOperationsException.java]: ??

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 16]
```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+			throw new GroupOperationsException(ex);
+		}
+	}
+
+	@Override
+	public Group createGroupUnion(PerunSession sess, Group resultGroup, Group operandGroup, boolean parentFlag) throws GroupOperationsException, InternalErrorException {
+		
+		//check if any of the groups is members group
+		if(resultGroup.getName().equals(VosManager.MEMBERS_GROUP) || operandGroup.getName().equals(VosManager.MEMBERS_GROUP)) {
+			throw new InternalErrorException("Union cannot be created on members group.");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Unless there is another check which prevent user to call this method with MEMBERS_GROUP in parameter, another exception should be used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 17]
```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+		}
+
+		// check if both groups are from same VO
+		if (resultGroup.getVoId() != operandGroup.getVoId()) {
+			throw new InternalErrorException("Groups are not from same VO");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Unless there is another check which prevent user this condition, another exception should be used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 18]
```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+		}
+
+		// check if result group is the same as operand group
+		if (resultGroup.getId() == operandGroup.getId()) {
+			throw new InternalErrorException("Result group cannot be the same as operand group.");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Unless there is another check which prevent user this condition, another exception should be used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 19]
```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+		} 
+
+		// check if there is already a record of these two groups
+		if (this.groupsManagerImpl.isRelationBetweenGroups(resultGroup, operandGroup)) {
+			throw new InternalErrorException("There is already an operation defined between these two groups: " + resultGroup + " and " + operandGroup 
+					+ " or they are in hierarchical structure.");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Unless there is another check which prevent user this condition, another exception should be used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 20]
```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+		}
+
+		// check cycle between groups
+		if (checkGroupsCycle(sess, resultGroup.getId(), operandGroup.getId())) {
+			throw new InternalErrorException("This relation would create group transitivity.");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Unless there is another check which prevent user this condition, another exception should be used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java [hunk 21]
```diff
@@ -2060,4 +2179,108 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
...
+		// do the operation logic
+		processRelationMembers(sess, resultGroup, getGroupMembers(sess, operandGroup), operandGroup.getId(), true);
+
+		return resultGroup;
+	}
+
+	@Override
+	public void removeGroupUnion(PerunSession sess, Group resultGroup, Group operandGroup, boolean parentFlag) throws GroupOperationsException, InternalErrorException {
+		if (!groupsManagerImpl.isOneWayRelationBetweenGroups(resultGroup, operandGroup)) {
+			throw new InternalErrorException("Relation does not exist between result group " + resultGroup + " and operand group" + operandGroup + ".");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java]: Unless there is another check which prevent user this condition, another exception should be used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java [hunk 22]
```diff
@@ -279,16 +281,17 @@ public boolean isUserMemberOfGroup(PerunSession sess, User user, Group group) th
 
 	public List<Member> getGroupMembers(PerunSession sess, Group group) throws InternalErrorException {
 		try {
-			return jdbc.query("select " + MembersManagerImpl.memberMappingSelectQuery + ", groups_members.membership_type as membership_type from groups_members join members on members.id=groups_members.member_id " +
-					" where groups_members.group_id=?", MembersManagerImpl.MEMBER_MAPPER, group.getId());
+			return jdbc.query("select " + MembersManagerImpl.memberMappingSelectQuery + ", groups_members.membership_type as membership_type,  " +
+					"groups_members.source_group_id as source_group_id from groups_members join members on members.id=groups_members.member_id " +
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java]: Guess the MembersManagerImpl.memberMappingSelectQuery could be expanded with ", groups_members.membership_type as membership_type, ..."

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java [hunk 23]
```diff
@@ -657,4 +666,51 @@ public int getGroupsCount(PerunSession sess) throws InternalErrorException {
...
+	@Override
+	public List<Integer> getRelatedGroupsIds(PerunSession sess, int groupId) {
+		return jdbc.queryForList("SELECT result_gid FROM groups_groups WHERE operand_gid=" + groupId, Integer.class);
+	}
+
+	@Override
+	public void removeGroupUnion(PerunSession sess, Group resultGroup, Group operandGroup) throws InternalErrorException {
+		if (0 == jdbc.update("DELETE FROM groups_groups WHERE result_gid = ? AND operand_gid = ?",
+				resultGroup.getId(), operandGroup.getId())) {
+			throw new InternalErrorException("There is no relation between " + resultGroup + " and " + operandGroup);
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java]: Use other exception.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java [hunk 24]
```diff
@@ -657,4 +666,51 @@ public int getGroupsCount(PerunSession sess) throws InternalErrorException {
...
+
+	@Override
+	public boolean isOneWayRelationBetweenGroups(Group resultGroup, Group operandGroup) {
+		return 1 <= jdbc.queryForInt("SELECT count(1) FROM groups_groups WHERE result_gid = ? AND operand_gid = ?",
+				resultGroup.getId(), operandGroup.getId());
+	}
+
+	@Override
+	public List<Integer> getGroupRelations(PerunSession sess, int groupId) {
+		return jdbc.queryForList("SELECT result_gid FROM groups_groups WHERE operand_gid=" + groupId, Integer.class);
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/GroupsManagerImpl.java]: :poop: Always use prepared statements (Substitute variables for '?' in SQL query)!

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/MembersManagerImpl.java [hunk 25]
```diff
@@ -49,9 +49,25 @@ public Member mapRow(ResultSet rs, int i) throws SQLException {
...
+	};
+
+	protected static final RowMapper<Member> GROUPS_MEMBER_MAPPER = new RowMapper<Member>() {
+		public Member mapRow(ResultSet rs, int i) throws SQLException {
+			Member member = new Member(rs.getInt("members_id"), rs.getInt("members_user_id"), rs.getInt("members_vo_id"), Status.getStatus(rs.getInt("members_status")),
+					rs.getString("members_created_at"), rs.getString("members_created_by"), rs.getString("members_modified_at"), rs.getString("members_modified_by"),
+					rs.getInt("members_created_by_uid") == 0 ? null : rs.getInt("members_created_by_uid"),
+					rs.getInt("members_modified_by_uid") == 0 ? null : rs.getInt("members_modified_by_uid"));
+			member.setSourceGroupId(rs.getInt("source_group_id"));
+			member.setMembershipType(MembershipType.getMembershipType(rs.getInt("membership_type")));
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/MembersManagerImpl.java]: Isn't there risk of SQL exception like on line 56?

**Comment by stavamichal** [comment]
Already deployed.

**Pull request closed without merging**